### ISSUE TITLE
Attribute to apply transparent wrappers to fields

### ISSUE THREAD

**Issue by djkoloski** [djkoloski/rkyv#122]
Most rkyv functionality can be implemented in terms of wrapper types around some data. For example:

```rust
pub struct Compress<T>(T);

// implement archive, serialize, and deserialize

#[derive(Archive, Serialize, Deserialize)]
pub struct Example {
    name: Compress<String>,
}
```

This is a great way to take advantage of type extensions to quickly leverage other libraries. However, we're still left with a `Wrapper<T>` which is just noise for the unarchived field. A better option would be to have the derive macros automatically perform this transformation for us. This would ideally lead to some syntax like:

```rust
#[derive(Archive, Serialize, Deserialize)]
pub struct Example {
    #[with(Compress)]
    name: String,
}
```

that would wrap the field `&T` in a `Wrapper<T>` and archive the `&Wrapper<&T>`. During deserialization, it would deserialize it into the original type. Applying multiple wrappers should also be easy:

```rust
#[derive(Archive, Serialize, Deserialize)]
pub struct Example {
    #[with(Intern, Compress)]
    name: String,
}
```

with wrappers being applied left to right (i.e. `Intern<&Compress<&T>>`).

Wrapper types should already be expressible using the traits currently available:

```rust
pub struct Wrap<'a, T>(&'a T);

pub struct ArchivedWrap<T>(T);

pub struct WrapResolver<T: Archive>(T::Resolver);

impl<'a, T> Archive for Wrap<'a, T> {
    type Archived = ArchivedWrap<T::Archived>;
    type Resolver = WrapResolver<T>;

    fn resolve(&self, pos: usize, resolver: Self::Resolver, out: &mut MaybeUninit<Self::Archived>) {
        self.0.resolve(pos, resolver.0, project_struct!(out: Self::Archived => 0);
    }
}

impl<'a, T: Serialize<S>, S: Fallible + ?Sized> Serialize<S> for Wrap<'a, T> {
    fn serialize(&self, serializer: &mut S) -> Result<Self::Resolver, S::Error> {
        Ok(WrapResolver(self.0.serialize(serializer)?))
    }
}

impl<'a, T: Archive, D: Fallible + ?Sized> Deserialize<T, D> for ArchivedWrap<T::Archived>
where
    T::Archived: Deserialize<T, D>,
{
    fn deserialize(&self, deserializer: &mut D) -> Result<T, D::Error> {
        Ok(self.0.deserialize(deserializer))
    }
}
```

but the bound `T: Archive<Archived = Self>` may need to be lifted from `Deserialize`.

**Comment by djkoloski** [comment]
Need to make sure we can support enum variants, e.g.

```rust
enum Segment<&'a T> {
    Primary(&'a T),
    Secondary(&'a T),
}

// wrapper implementation

#[derive(Archive, Serialize, Deserialize)]
pub struct Example {
    #[with(Segment::Secondary)]
    name: String,
}
```

I think it should be okay as long as the arguments are treated as constructors.

**Comment by djkoloski** [comment]
Fixed by cd320234423b076b3edd7c29e31eb66964e02377
